File: designpattern/src/main/java/creational/singleton/Singleton.java
package creational.singleton;

public class Singleton {

    private static Singleton instance;
    private int id = 0;

    private Singleton() {
    }

    public static Singleton getInstance() {
        if(null == instance) {
            synchronized (Singleton.class) {
                if (null == instance) {
                    instance = new Singleton();
                }
            }
        }
        return instance;
    }

    public int getUID() {
        return id++;
    }

    public static void main(String[] args) {
        Singleton o1 = Singleton.getInstance();
        Singleton o2 = Singleton.getInstance();

        System.out.println(o1.getUID());
        System.out.println(o2.getUID());
    }
}

File: designpattern/src/main/cpp/creational/singleton/singleton.cpp
#include "singleton.h"

#include <iostream>
#include <mutex>
using std::mutex;

mutex lock;
singleton *singleton::instance = NULL;
//int singleton::id = 0;
singleton::singleton() {
    id = 0;
}

singleton::~singleton() {

}

singleton *singleton::getInstance() {

    if(NULL == instance) {
        lock.lock();
        if(NULL == instance) {
            instance = new singleton;
        }
        lock.unlock();
    }
    return instance;
}

void singleton::release() {
        if(NULL != instance) {
        lock.lock();
        if(NULL != instance) {
            delete instance;
        }
        lock.unlock();
    }
}

int singleton::getUID() {
    lock.lock();
    ++id;
    lock.unlock();
    return id;
}